documentation/dataplane/dash-flow-api.md

```diff
@@ -0,0 +1,633 @@
+  
+    sai_flow_entry_t *flow_entry;
+  
+    uint32_t attr_count; 
+  
+    sai_attribute_t *attr_list; 
+
+} sai_flow_bulk_get_session_event_data_t;
+
+/**
+ * @brief dash flow get bulk session notification
+ *
+ * Passed as a parameter into sai_initialize_switch()
+ *
+ * @count data[count]
+ *
+ * @param[in] count Number of notifications
+ * @param[in] data Array of flow bulk get session events
+ */
+typedef void (*sai_flow_bulk_get_session_event_notification_fn)(
+        _In_ uint32_t count,
+        _In_ const sai_flow_bulk_get_session_event_data_t *flow_bulk_get_session_event_data);
+```
+
+### Protobuf-based flow programming
+
+In addition to the flow state attributes, the flow state can be represented using protobuf for high-efficiency. The attribute of the flow entry is SAI_FLOW_ENTRY_ATTR_FLOW_DATA_PB.
+
+Although the content of both attributes and protobuf may be identical, their applications differ. Attributes enable incremental updates to individual properties, whereas protobuf necessitates a complete update.
+
+```protobuf
+syntax = "proto3";
+
+message SaiDashFlowKey {
+  uint64 eni_mac = 1;
+  string src_addr = 2;
+  string dst_addr = 3;
+  uint32 src_port = 4;
+  uint32 dst_port = 5;
+}
+
+message SaiDashFlowState {
+  uint32 version = 1; // SAI_FLOW_ENTRY_ATTR_VERSION
+  uint32 dash_flow_action = 2; // SAI_FLOW_ENTRY_ATTR_DASH_FLOW_ACTION
+  uint32 meter_class = 3; // SAI_FLOW_ENTRY_ATTR_METER_CLASS
+  bool is_bidirectional_flow = 4; // SAI_FLOW_ENTRY_ATTR_IS_BIDIRECTIONAL_FLOW
+  uint32 underlay_vni = 5; // SAI_FLOW_ENTRY_ATTR_UNDERLAY_VNI
+  string underlay_sip = 6; // SAI_FLOW_ENTRY_ATTR_UNDERLAY_SIP
+  string underlay_dip = 7; // SAI_FLOW_ENTRY_ATTR_UNDERLAY_DIP
+  string underlay_smac = 8; // SAI_FLOW_ENTRY_ATTR_UNDERLAY_SMAC
+  string underlay_dmac = 9; // SAI_FLOW_ENTRY_ATTR_UNDERLAY_DMAC
+  uint32 underlay2_vni = 10; // SAI_FLOW_ENTRY_ATTR_UNDERLAY2_VNI
+  string underlay2_sip = 11; // SAI_FLOW_ENTRY_ATTR_UNDERLAY2_SIP
+  string underlay2_dip = 12; // SAI_FLOW_ENTRY_ATTR_UNDERLAY2_DIP
+  string underlay2_smac = 13; // SAI_FLOW_ENTRY_ATTR_UNDERLAY2_SMAC
+  string underlay2_dmac = 14; // SAI_FLOW_ENTRY_ATTR_UNDERLAY2_DMAC
+  string dst_mac = 15; // SAI_FLOW_ENTRY_ATTR_DEST_MAC
+  string sip = 16; // SAI_FLOW_ENTRY_ATTR_SIP
+  string dip = 17; // SAI_FLOW_ENTRY_ATTR_DIP
+  string sip_mask = 18; // SAI_FLOW_ENTRY_ATTR_SIP_MASK
+  string dip_mask = 19; // SAI_FLOW_ENTRY_ATTR_DIP_MASK
+}
+
+message SaiDashFlowEntry {
+  SaiDashFlowKey flow_key = 1;
+  SaiDashFlowKey reverse_flow_key = 2;
+  SaiDashFlowState flow_state = 3;
+}
+```
+
+### Capability
+
+| Attribute Name                                     | Type           | Description                                        |
+| -------------------------------------------------- | -------------- | -------------------------------------------------- |
+| SAI_SWITCH_ATTR_DASH_CAPS_MAX_FLOW_TABLE_COUNT     | `sai_uint32_t` | The max number of flow tables that can be created  |
```

Comment:
since the different keys can be enabled, maybe we should add a capability to get which keys are supported.

Reply:
Done